src/components/pushpin-code/Changes.tsx

```diff
@@ -0,0 +1,172 @@
+      if (!doc) {
+        return
+      }
+
+      let lastSeenHeadsByDocUrl = workspaceDoc.persistedLastSeenHeads
+      if (!lastSeenHeadsByDocUrl) {
+        workspaceDoc.persistedLastSeenHeads = {}
+        // need to read persistedLastSeenHeads from workspaceDoc again, so we get the proxied automerge object
+        lastSeenHeadsByDocUrl = workspaceDoc.persistedLastSeenHeads
+      }
+
+      lastSeenHeadsByDocUrl[docUrl] = getHeads(doc)
+    })
+  }, [doc, changeWorkspaceDoc])
+}
+
+const CURRENTLY_VIEWED_DOC_URLS: { [url: PushpinUrl]: boolean } = {}
+
+function isDocUrlCurrentlyViewed(url: PushpinUrl): boolean {
+  return CURRENTLY_VIEWED_DOC_URLS[url] ?? false
+}
+
+export function useAutoAdvanceLastSeenHeads(docUrl: PushpinUrl) {
+  const [doc] = useDocument(parseDocumentLink(docUrl).documentId)
+  const advanceLastSeenHeads = useAdvanceLastSeenHeads(docUrl)
+
+  useEffect(() => {
+    CURRENTLY_VIEWED_DOC_URLS[docUrl] = true
+
+    return () => {
+      delete CURRENTLY_VIEWED_DOC_URLS[docUrl]
+    }
+  }, [docUrl])
+
+  useEffect(() => {
+    advanceLastSeenHeads()
+  }, [doc])
+}
+
+export type LastSeenHeads = Heads | "latestHeads"
+
+export interface LastSeenHeadsMap {
+  [url: PushpinUrl]: LastSeenHeads
+}
+
+export interface PersistedLastSeenHeadsMap {
+  [url: PushpinUrl]: Heads
+}
+
+export function useLastSeenHeads(
+  docUrl: PushpinUrl
+): LastSeenHeads | undefined {
+  const workspaceId = useContext(WorkspaceContext)
+  const [workspaceDoc, changeWorkspaceDoc] =
+    useDocument<WorkspaceDoc>(workspaceId)
+
+  if (CURRENTLY_VIEWED_DOC_URLS[docUrl]) {
+    return "latestHeads"
+  }
+
+  if (!workspaceDoc || !workspaceDoc.persistedLastSeenHeads) {
+    return undefined
+  }
+
+  const lastSeenHeads = workspaceDoc.persistedLastSeenHeads[docUrl]
+
+  if (!lastSeenHeads) {
+    changeWorkspaceDoc((workspaceDoc) => {
+      let lastSeenHeadsByDocUrl = workspaceDoc.persistedLastSeenHeads
+      if (!lastSeenHeadsByDocUrl) {
+        workspaceDoc.persistedLastSeenHeads = {}
+        // need to read persistedLastSeenHeads from workspaceDoc again, so we get the proxied automerge object
+        lastSeenHeadsByDocUrl = workspaceDoc.persistedLastSeenHeads
+      }
+
+      lastSeenHeadsByDocUrl[docUrl] = []
+    })
+  }
+
+  return lastSeenHeads
+}
+
+export function getUnseenPatches(
+  doc: Doc<any>,
+  lastSeenHeads?: LastSeenHeads
+): Patch[] {
+  const patches: Patch[] = []
+
+  if (!hasDocUnseenChanges(doc, lastSeenHeads)) {
+    return []
+  }
+
+  try {
+    const oldDoc = clone(view(doc, lastSeenHeads as Heads))
+
+    applyChanges(oldDoc, getChanges(oldDoc, doc), {
+      patchCallback: (patch: Patch) => {
+        patches.push(patch)
+      },
+    })
+
+    return patches
+  } catch (err) {
+    console.error(err)
+    return []
+  }
+}
+
+export function hasDocUnseenChanges(
+  doc: Doc<any>,
+  lastSeenHeads?: LastSeenHeads
+): boolean {
+  if (!lastSeenHeads || lastSeenHeads === "latestHeads") {
```

Comment:
I see what you mean, this *is* kind of gross. I haven't got a better idea right now.

More importantly, we should talk about what to do if there are no `lastSeenHeads`. Looking at how Discord behaves when you join a new server, all channels begin as "caught up" but as new messages arrive post-joining the "unread status" activates.